docs/writing_tests.md

```diff
@@ -0,0 +1,156 @@
+
+> Before adding a test case, consider whether there already exists a category it would fit into, or whether a new category is required.
+
+> The CTS supports disabling the compilation of entire categories for certain SYCL implementations.
+> See [Procedures](procedures.md) for more information.
+
+## Setting up a Simple Category & Test Case
+
+To create a new test category create the following files inside the [`tests`](../tests) directory:
+
+```
+tests
+└── simple
+   ├── CMakeLists.txt
+   └── simple.cpp
+```
+
+In `tests/simple/CMakeLists.txt` add the following boilerplate:
+
+```
+file(GLOB test_cases_list *.cpp)
+add_cts_test(${test_cases_list})
+```
+
+Then in `tests/simple/simple.cpp` add the following:
+
+```c++
+#include "../common/common.h"
+
+TEST_CASE("a simple test case", "[simple]") {
+    sycl::buffer<int> buf(1);
+    sycl::queue queue = sycl_cts::util::get_cts_object::queue();
+    queue.submit([&](sycl::handler& cgh) {
+        sycl::accessor w{buf, cgh, sycl::write_only};
+        cgh.single_task<class simple_kernel>([=] {
+            w[0] = 42;
+        });
+    });
+
+    sycl::host_accessor r{buf, sycl::read_only};
+    CHECK(r[0] == 42);
+}
+```
+
+This adds a test case with the description `"a simple test case"` and the *tag* `[simple]`.
+Both can later be used to narrow down the set of test cases that will be executed during runtime.
+
+When configuring CMake, the new test category will automatically be detected and a target with the name `test_simple` is added.
+You can run the test case by either executing `./bin/test_simple` directly, or alternatively as part of `./bin/test_all`.
+
+> Note that for historic reasons, the CTS currently contains many test cases that are written in a different style.
+> Please see [New-style vs Legacy Test Cases](#new-style-vs-legacy-test-cases) for more information.
+
+## Important Catch2 Concepts
+
+The SYCL CTS relies on [Catch2](https://github.com/catchorg/Catch2/) as its underlying testing framework.
+This section will list the most important concepts required to write tests with Catch2.
+For a comprehensive overview of all features, please refer to the [Catch2 documentation](https://github.com/catchorg/Catch2/tree/devel/docs).
+In addition, the CTS provides several custom utilities to extend Catch2's feature set.
+See [Special Macros & Custom Matchers](#special-macros--custom-matchers) for more information.
+
+### Test Case Macros
+
+Catch2 provides several macros of varying complexity for defining test cases.
+While different macros take different parameters, they all require a *description* and optionally a list of *tags* to be specified.
+
+
+- `TEST_CASE`: This is the most basic macro, useful for test cases that do not require multiple compile-time definitions.
+- `TEMPLATE_TEST_CASE`
+- `TEMPLATE_TEST_CASE_SIG`
+
+**TODO**
+
+### Sections
+
+**TODO**
+
+### Generators
+
+**TODO**
+
+## Special Macros & Custom Matchers
+
+**TODO**
```

Comment:
I don't understand why do we need to describe `cathc2` macros in our doc. Can't we just leave a link to the `catch2` documentation? 

Also, I think it is more important to describe macros that were implemented only in our repo. For example `DISABLE_FOR_TEST_CASE` and `DISABLE_FOR_TEMPLATE_TEST_CASE_SIG`

Reply:
I completely agree, it was not my intent to go to great lengths explaining Catch2 here. Instead I just wanted to list the most important macros and how they tend to be used in the CTS (for example, using `TEMPLATE_TEST_CASE_SIG` for testing APIs with 1/2/3 dimensions).

> Also, I think it is more important to describe macros that were implemented only in our repo. For example DISABLE_FOR_TEST_CASE and DISABLE_FOR_TEMPLATE_TEST_CASE_SIG

Yes, this is what the "Special Macros & Custom Matchers" was intended for!